### PR TITLE
[2chain] add 2-chain related Timeout structs and implement safety rules

### PR DESCRIPTION
**Pull request by zekun000**
Add 2chain related Timeout structure which carries the highest quorum cert. Test added for verification.

Implement corresponding safety rules in safety_rules_2chain and add tests.

Remaining work:
- Add new Timeout in Vote
- Add new TimeoutCert in SyncInfo
- Add gated codepath for using 2-chain protocol
- Add on-chain config to switch the protocol

## Motivation

Add 2chain related Timeout structure which carries the highest quorum cert. Test added for verification.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

Added tests.

## Related PRs

(If this PR adds or changes functionality, please take some time to update or suggest changes to the docs at https://developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

**Review by sasha8 — DISMISSED** [review]
I have reviewed timeout_2chain.rs and safety_rules_2chain.rs and the logic looks good to me.

File: consensus/safety-rules/src/safety_rules_2chain.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        if timeout.round() < safety_data.last_voted_round {
+            return Err(Error::IncorrectLastVotedRound(
+                timeout.round(),
+                safety_data.last_voted_round,
+            ));
+        }
+        if timeout.round() > safety_data.last_voted_round {
+            self.verify_and_update_last_vote_round(timeout.round(), &mut safety_data)?;
+            self.persistent_storage.set_safety_data(safety_data)?;
+        }
```
Comment (aching) [inline consensus/safety-rules/src/safety_rules_2chain.rs:40]: It looks like a malicious caller can set a timeout.round() that is arbitrarily larger than safety_data.last_voted_round and advance the safety data round arbitrarily high (liveness attack).

Shouldn't only the TC, QC, or a vote be able to advance the safety data round?

I do note this can happen in the original sign_timeout() as well.

File: consensus/safety-rules/src/serializer.rs [hunk 2]
```diff
@@ -33,19 +45,31 @@ impl SerializerService {
...
+                SafetyRulesInput::ConstructAndSignVote(vote_proposal) => {
+                    bcs::to_bytes(&self.internal.construct_and_sign_vote(&vote_proposal))
+                }
+                SafetyRulesInput::SignProposal(block_data) => {
+                    bcs::to_bytes(&self.internal.sign_proposal(*block_data))
+                }
+                SafetyRulesInput::SignTimeout(timeout) => {
+                    bcs::to_bytes(&self.internal.sign_timeout(&timeout))
+                }
+                SafetyRulesInput::SignTimeoutWithQC(timeout, maybe_tc) => bcs::to_bytes(
```
Comment (aching) [inline consensus/safety-rules/src/serializer.rs:61]: Maybe add a comment about the below messages only for two-chain?

File: consensus/safety-rules/src/test_utils.rs [hunk 3]
```diff
@@ -66,7 +66,7 @@ pub fn make_proposal_with_qc_and_proof(
         Block::new_proposal(
             payload,
             round,
-            duration_since_epoch().as_secs(),
+            qc.certified_block().timestamp_usecs() + 1,
```
Comment (aching) [inline consensus/safety-rules/src/test_utils.rs:69]: Why is this change necessary?

File: consensus/safety-rules/src/tests/suite.rs [hunk 4]
```diff
@@ -660,7 +666,6 @@ fn test_validator_not_in_set(safety_rules: &Callback) {
         Some(next_epoch_state),
         key.as_ref(),
     );
-    safety_rules.construct_and_sign_vote(&a2).unwrap();
```
Comment (aching) [inline consensus/safety-rules/src/tests/suite.rs:663]: Just some random cleanup?

File: consensus/safety-rules/src/tests/suite.rs [hunk 5]
```diff
@@ -708,7 +713,6 @@ fn test_reconcile_key(_safety_rules: &Callback) {
         Some(next_epoch_state),
         None,
     );
-    safety_rules.construct_and_sign_vote(&a2).unwrap();
```
Comment (aching) [inline consensus/safety-rules/src/tests/suite.rs:711]: Just some random cleanup?

**Review by aching — COMMENTED** [review]
Looks like a good start on the 2-chain protocol!

Currently the description reads, "Add 2chain related Timeout structure which carries the highest quorum cert. Test added for verification.
Implement corresponding safety rules in safety_rules_2chain and add tests."

Can you also please describe what does not work and what is the plan for implementing the remaining work?

File: consensus/safety-rules/src/safety_rules_2chain.rs (hunk 1 above)
Comment (zekun000) [reply consensus/safety-rules/src/safety_rules_2chain.rs:40]: advancing timeout round arbitrarily is actually a safety attack (with lower round qc), that's why we have safe_to_timeout check first

File: consensus/safety-rules/src/test_utils.rs (hunk 3 above)
Comment (zekun000) [reply consensus/safety-rules/src/test_utils.rs:69]: because I added missing check for `block.verify_well_formed` and it requires strict increasing timestamps

**Comment by aching** [comment]
Also, while there is some separation in the code for 2-chain (e.g. safety_rules_2chain.rs), it would be great to be as explicit as possible to which parts of the code are for 2-chain only to avoid confusion from any readers/editors of the code during the transition.

File: consensus/safety-rules/src/tests/suite.rs (hunk 4 above)
Comment (zekun000) [reply consensus/safety-rules/src/tests/suite.rs:663]: no, this fails after we added block.verify_well_formed check

File: consensus/safety-rules/src/tests/suite.rs (hunk 5 above)
Comment (zekun000) [reply consensus/safety-rules/src/tests/suite.rs:711]: ditto

File: consensus/safety-rules/src/safety_rules.rs [hunk 6]
```diff
@@ -32,15 +33,18 @@ use diem_types::{
...
-    persistent_storage: PersistentSafetyStorage,
-    execution_public_key: Option<Ed25519PublicKey>,
-    export_consensus_key: bool,
-    validator_signer: Option<ConfigurableValidatorSigner>,
-    epoch_state: Option<EpochState>,
+    pub(crate) persistent_storage: PersistentSafetyStorage,
+    pub(crate) execution_public_key: Option<Ed25519PublicKey>,
+    pub(crate) export_consensus_key: bool,
+    pub(crate) validator_signer: Option<ConfigurableValidatorSigner>,
+    pub(crate) epoch_state: Option<EpochState>,
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/safety-rules/src/safety_rules.rs:47]: pub(crate) is mainly for the separate file which I think it's better for review. we can merge it back and remove it bu it's not high pri

File: consensus/consensus-types/src/safety_data.rs [hunk 7]
```diff
@@ -10,7 +10,11 @@ use std::fmt;
 pub struct SafetyData {
     pub epoch: u64,
     pub last_voted_round: u64,
+    // highest 2-chain round, used for 3-chain
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/consensus-types/src/safety_data.rs:13]: then we need to update the text after we enable it

File: consensus/safety-rules/src/serializer.rs (hunk 2 above)
Comment (zekun000) [reply consensus/safety-rules/src/serializer.rs:61]: the comment is on the public api (TSafetyRules)

File: consensus/consensus-types/src/safety_data.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (aching) [reply consensus/consensus-types/src/safety_data.rs:13]: Fair enough.

[the remark this answers is not in the document]
Comment (aching) [reply consensus/consensus-types/src/safety_data.rs:13]: Still I think it's a bit confusing for developers who look at the code now and don't understand the differences in the protocol and the implementations.  I think we should address this in future PRs.

File: consensus/safety-rules/src/safety_rules_2chain.rs (hunk 1 above)
Comment (aching) [reply consensus/safety-rules/src/safety_rules_2chain.rs:40]: Got it.  Not a part of this PR, but what about sign_timeout()?

**Review by aching — DISMISSED** [review]

**Comment by aching** [comment]
FYI, we need to open up a standards DIP for the upcoming 2-chain protocol and gather more feedback from the community.  Can you please do that @zekun000 ?

File: consensus/safety-rules/src/safety_rules_2chain.rs (hunk 1 above)
Comment (zekun000) [reply consensus/safety-rules/src/safety_rules_2chain.rs:40]: that's only about liveness and doesn't seem matter much

Comment (aching) [reply consensus/safety-rules/src/safety_rules_2chain.rs:40]: I think it's not good from a TCB perspective that a corrupted validator can hurt liveness in the 3-chain protocol.  That being said, if we upgrade to 2-chain soon, perhaps this isn't as necessary to fix.
